src/merkle/mmr/accumulator.rs

```diff
@@ -35,25 +31,57 @@ pub struct MmrPeaks {
 impl MmrPeaks {
     /// Hashes the peaks.
     ///
-    /// The hashing is optimized to work with the Miden VM, the procedure will:
+    /// The procedure will:
+    /// - Flatten and pad the peaks to a vector of Felts.
+    /// - Hash the vector of Felts.
+    pub fn hash_peaks(&self) -> Word {
+        Rpo256::hash_elements(&self.flatten_and_pad_peaks()).into()
+    }
+
+    pub fn verify(&self, value: Word, opening: MmrProof) -> bool {
+        let root = &self.peaks[opening.peak_index()];
+        opening.merkle_path.verify(opening.relative_pos() as u64, value, root)
+    }
+
+    /// Flattens and pads the peaks to make hashing inside of the Miden VM easier.
     ///
+    /// The procedure will:
+    /// - Flatten the vector of Words into a vector of Felts.
     /// - Pad the peaks with ZERO to an even number of words, this removes the need to handle RPO padding.
     /// - Pad the peaks to a minimum length of 16 words, which reduces the constant cost of
     ///   hashing.
-    pub fn hash_peaks(&self) -> Word {
-        let mut copy = self.peaks.clone();
+    fn flatten_and_pad_peaks(&self) -> Vec<Felt> {
+        let num_peaks = self.peaks.len();
 
-        if copy.len() < 16 {
-            copy.resize(16, [ZERO; WORD_SIZE])
-        } else if copy.len() % 2 == 1 {
-            copy.push([ZERO; WORD_SIZE])
-        }
+        // To achieve the padding rules above we calculate the length of the final vector.
+        let len = if num_peaks < 16 {
+            64
+        } else if num_peaks % 2 == 1 {
+            (num_peaks + 1) * 4
+        } else {
+            num_peaks * 4
+        };
 
-        Rpo256::hash_elements(&copy.as_slice().concat()).into()
+        let mut elements = Vec::with_capacity(len);
+        elements.extend_from_slice(&self.peaks.as_slice().concat());
+        elements.resize(len, ZERO);
+        elements
     }
+}
 
-    pub fn verify(&self, value: Word, opening: MmrProof) -> bool {
-        let root = &self.peaks[opening.peak_index()];
-        opening.merkle_path.verify(opening.relative_pos() as u64, value, root)
+/// Converts the MMR peaks into a vector of Felts.
+///
+/// The vector of Felts is arranged as follows:
+/// [num_leaves, PEAK_0, PEAK_1, ..., PEAK_N]
+///
+/// num_leaves is contains the number of leaves in the MMR.
+/// PEAK_0 is the 0'th peak in the MMR.
+/// PEAK_1 is the 1'st peak in the MMR.
+/// PEAK_N is the N'th peak in the MMR.
+impl From<&MmrPeaks> for Vec<Felt> {
+    fn from(mmr: &MmrPeaks) -> Self {
+        let mut elements = mmr.flatten_and_pad_peaks();
+        elements.insert(0, Felt::new(mmr.num_leaves as u64));
```

Comment:
note: the MMR uses a word for the `Felt`, not a single element. This needs 3 padding zeros.

Reply:
noted - thanks.  In the process of implementing this now.